tests/DateScalarTest.php

```diff
@@ -0,0 +1,130 @@
+
+abstract class DateScalarTest extends TestCase
+{
+    /**
+     * @dataProvider invalidDateValues
+     *
+     * @param mixed $value An invalid value for a date
+     */
+    public function testThrowsIfSerializingInvalidDates($value): void
+    {
+        $this->expectException(InvariantViolation::class);
+
+        $this->scalarInstance()->serialize($value);
+    }
+
+    /**
+     * @dataProvider invalidDateValues
+     *
+     * @param mixed $value An invalid value for a date
+     */
+    public function testThrowsIfParseValueInvalidDate($value): void
+    {
+        $this->expectException(Error::class);
+
+        $this->scalarInstance()->parseValue($value);
+    }
+
+    /**
+     * Those values should fail passing as a date.
+     *
+     * @return Generator<array-key, array{mixed}>
+     */
+    public function invalidDateValues(): Generator
+    {
+        yield [1];
+        yield ['rofl'];
+        yield [null];
+        yield [''];
+        yield ['2020-02-30'];
+        yield ['2020-02-30 01:02:03'];
+        yield ['2017-01-001T00:00:00Z'];
+        yield ['2017-02-31T00:00:00Z'];
+        yield ['2018-11-31T00:00:00Z'];
+        yield ['2019-02-29T00:00:00Z'];
+        yield ['2020-02-30T00:00:00Z'];
+    }
+
+    /**
+     * @dataProvider validDates
+     */
+    public function testParsesValueString(string $value, string $expected): void
+    {
+        $parsedValue = $this->scalarInstance()->parseValue($value);
+
+        self::assertSame($expected, $parsedValue->format('Y-m-d\TH:i:s.uP'));
+    }
+
+    /**
+     * @dataProvider validDates
+     */
+    public function testParsesLiteral(string $value, string $expected): void
+    {
+        $dateLiteral = new StringValueNode(
+            ['value' => $value]
+        );
+        $parsed = $this->scalarInstance()->parseLiteral($dateLiteral);
+
+        self::assertSame($expected, $parsed->format('Y-m-d\TH:i:s.uP'));
+    }
+
+    public function testThrowsIfParseLiteralNonString(): void
+    {
+        $this->expectException(Error::class);
+
+        $this->scalarInstance()->parseLiteral(
+            new IntValueNode([])
+        );
+    }
+
+    public function testSerializesDateTimeInterfaceInstance(): void
+    {
+        $now = new DateTimeImmutable();
+        $result = $this->scalarInstance()->serialize($now);
+
+        self::assertNotEmpty($result);
+    }
+
+    /**
+     * @dataProvider canonicalizeDates
+     */
+    public function testCanonicalizesValidDateString(string $date, string $canonical): void
+    {
+        $result = $this->scalarInstance()->serialize($date);
+
+        self::assertSame($canonical, $result);
+    }
+
+    /**
+     * The specific instance under test.
+     */
+    abstract protected function scalarInstance(): DateScalar;
+
+    /**
+     * Data provider for valid date strings and expected dates.
+     *
+     * @return Generator<array-key, array{string, string}>
+     */
+    abstract public function validDates(): Generator;
```

Comment:
```suggestion
     * @return iterable<array{string, string}>
     */
    abstract public function validDates(): iterable;
```

Reply:
I'm curious why 🤔

Reply:
It is simpler and places no unnecessary constraint on its children, they are free to use `yield`, return an `array` or return a `Traversable`.